File: src/utils/backupCodes.ts
import { customAlphabet } from 'nanoid';
import { createHash } from 'crypto';

// Use a custom alphabet for backup codes that's unambiguous
const BACKUP_CODE_ALPHABET = '23456789ABCDEFGHJKLMNPQRSTUVWXYZ';
const BACKUP_CODE_LENGTH = 10;
const NUMBER_OF_CODES = 10;

export class BackupCodesManager {
  private static generateSingleCode(): string {
    const nanoid = customAlphabet(BACKUP_CODE_ALPHABET, BACKUP_CODE_LENGTH);
    return nanoid();
  }

  static generateCodes(): string[] {
    const codes: string[] = [];
    while (codes.length < NUMBER_OF_CODES) {
      const code = this.generateSingleCode();
      if (!codes.includes(code)) {
        codes.push(code);
      }
    }
    return codes;
  }

  static hashCode(code: string): string {
    return createHash('sha256')
      .update(code)
      .digest('hex');
  }

  static formatCode(code: string): string {
    // Format code into groups of 5 characters
    return code.match(/.{1,5}/g)?.join('-') || code;
  }

  static validateCode(code: string, hashedCodes: string[]): boolean {
    const normalizedCode = code.replace(/-/g, '').toUpperCase();
    const hashedCode = this.hashCode(normalizedCode);
    return hashedCodes.includes(hashedCode);
  }
}

export interface BackupCodesStorage {
  hashedCodes: string[];
  lastGenerated: string;
  usedCodes: string[];
}

export const storeBackupCodes = async (
  userId: string,
  codes: string[]
): Promise<void> => {
  const hashedCodes = codes.map(code => BackupCodesManager.hashCode(code));
  const storage: BackupCodesStorage = {
    hashedCodes,
    lastGenerated: new Date().toISOString(),
    usedCodes: [],
  };

  // Store in secure backend storage
  await fetch('/api/auth/backup-codes', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      userId,
      codes: storage,
    }),
  });
};

export const validateAndUseBackupCode = async (
  userId: string,
  code: string
): Promise<boolean> => {
  try {
    const response = await fetch('/api/auth/backup-codes/verify', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        userId,
        code,
      }),
    });

    if (!response.ok) {
      throw new Error('Failed to verify backup code');
    }

    const result = await response.json();
    return result.valid;
  } catch (error) {
    console.error('Backup code validation failed:', error);
    return false;
  }
}; 

File: src/utils/backupCodes.py
import secrets
import random
import hashlib
import requests
from typing import List
from datetime import datetime
from dataclasses import dataclass


BACKUP_CODE_ALPHABET = '23456789ABCDEFGHJKLMNPQRSTUVWXYZ'
BACKUP_CODE_LENGTH = 10
NUMBER_OF_CODES = 10


class BackupCodesManager:
    @staticmethod
    def __generate_single_code():
        return ''.join(secrets.choice(BACKUP_CODE_ALPHABET) for _ in range(BACKUP_CODE_LENGTH))

    @staticmethod
    def generate_codes():
        codes = []
        while len(codes) < NUMBER_OF_CODES:
            code = BackupCodesManager.__generate_single_code()
            if code not in codes:
                codes.append(code)
        return codes

    @staticmethod
    def hash_code(code: str):
        return hashlib.sha256(code.encode()).hexdigest()

    @staticmethod
    def format_code(code: str):
        return '-'.join(code[i:i + 5] for i in range(0, len(code), 5))

    @staticmethod
    def validate_code(code: str, hashed_codes: List[str]):
        normalized_code = code.replace('-', '').upper()
        hashed_code = BackupCodesManager.hash_code(normalized_code)
        return hashed_code in hashed_codes
    
@dataclass
class BackupCodesStorage:
    hashedCodes: List[str]
    lastGenerated: str
    usedCodes: List[str]
        

async def store_backup_codes(user_id: str, codes: List[str]):
    hashed_codes = [BackupCodesManager.hash_code(code) for code in codes]
    storage = BackupCodesStorage(
        hashedCodes=hashed_codes,
        lastGenerated=datetime.now().isoformat(),
        usedCodes=[],
    )

    # Store in secure backend storage (replace with actual implementation)
    response = requests.post(
        '/api/auth/backup-codes',
        headers={'Content-Type': 'application/json'},
        json={
            'userId': user_id,
            'codes': storage.__dict__,
        },
    )
    response.raise_for_status()  # Raise an exception for bad responses


async def validate_and_use_backup_code(user_id: str, code: str):
    try:
        response = requests.post(
            '/api/auth/backup-codes/verify',
            headers={'Content-Type': 'application/json'},
            json={'userId': user_id, 'code': code},
        )
        response.raise_for_status()  # Raise an exception for bad responses
        result = response.json()
        return result.get('valid', False)
    except requests.exceptions.RequestException as error:
        print('Backup code validation failed:', error)
        return False